src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/JsonCollectionConverter.cs

```diff
@@ -275,6 +278,9 @@ protected static JsonConverter<TElement> GetElementConverter(ref WriteStack stat
 
             ConvertCollection(ref state, options);
             value = (TCollection)state.Current.ReturnValue!;
+
+            jsonTypeInfo.OnDeserialized?.Invoke(value);
```

Comment:
If `TCollection` is a struct this would result in the value being boxed again and potentially breaking user callback that depend on the same reference being passed to the callback.
```suggestion
            object result = state.Current.ReturnValue!;
            jsonTypeInfo.OnDeserialized?.Invoke(result);
            value = (TCollection)result!;
```